// coreui/src/components/TextInput/TextInput.js
import React, { PropTypes } from 'react';
import Shared from '../../Shared';
import defaultTheme from '../../theme/components/TextInput';
import cx from 'classnames/dedupe';
import componentFromProp from 'recompose/componentFromProp';
import compose from 'recompose/compose';
import mapProps from 'recompose/mapProps';
import withHandlers from 'recompose/withHandlers';
import { merge, omit } from 'ramda';

const systemStyles = { };

const TextInputContainer = compose(
  withHandlers({
    onChange: (props) => ({ target }) =>
    props.onChange && props.onChange(target.value),
  }),
  mapProps(({ className, sheet, style, tagName, theme, ...rest }) => ({
    className: cx(sheet.classes.textInput, theme.classes.textInput, className),
    component: tagName,
    style: merge(theme.styles.textInput, style),
    ...omit(['errors', 'invalid'], rest),
  }))
)(componentFromProp('component'));

const StyledTextInput = Shared.injectSheet(systemStyles)(TextInputContainer);

const TextInput = (props) =>
  <StyledTextInput {...props}>{props.children}</StyledTextInput>;

const classes = defaultTheme.classes;
const options = defaultTheme.options;
const styles = defaultTheme.styles;

TextInput.defaultProps = { tagName: 'input', theme: { classes, options, styles } };

TextInput.displayName = 'TextInput';

TextInput.propTypes = {
  children: PropTypes.node,

  className: PropTypes.string,

  tagName: PropTypes.string,

  theme: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),

  type: PropTypes.string,
};

Shared.registerComponent('TextInput', TextInput);

export default TextInput;
